### DISCUSSION TITLE
What's the point of watchReadContract invoking callback on chain change?

### DISCUSSION THREAD

**Discussion by cilphex** [wevm/wagmi#1469] [Q&A] [answered]
What's the point of `watchReadContract` invoking callback on chain change? If the chain changes to another chain that your contract is deployed to, aren't you going to have to unwatch first, and then call `watchReadContract` again with new parameters, specifically a new contract address?

If you switch to a chain where your contract is not deployed at the originally specified address (or is not deployed to at all), this results in an error when the `watchReadContract` callback is called. This is not an error that is _passed to_ the callback; rather just a background error that seems un-catchable.

Further, `watchReadContract` has a [return value](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/actions/contracts/watchReadContracts.ts#L33-L36) that is not specified in [the docs](https://wagmi.sh/core/actions/watchReadContract).

**Comment by jxom** [comment, marked as answer]
> If the chain changes to another chain that your contract is deployed to, aren't you going to have to unwatch first, and then call watchReadContract again with new parameters, specifically a new contract address?

Some contracts share the same address across chains, so can make sense for those scenarios (without needing to unwatch/watch again).

> This is not an error that is passed to the callback; rather just a background error that seems un-catchable.

You can catch it by wrapping `watchReadContract` in a try/catch block.

> `watchReadContract` has a [return value](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/actions/contracts/watchReadContracts.ts#L33-L36) that is not specified in [the docs](https://wagmi.sh/core/actions/watchReadContract).

Thanks! We can update this.
